store.rb

```diff
@@ -118,4 +153,26 @@ def get_node(node)
       raise e
     end
   end
+
+  # immediately update cache if node joins/leaves
+  def watch()
+    return if @zk.nil?
+    @root_watcher ||= @zk.register("/", :only => :child) do |event|
+      @log.info "Children added/deleted"
+      reload_instances
+    end
+  end
+
+  def reload_instances()
+    instances = load_instances_from_zk()
+    cache_instances(instances)
+  end
+
+  def get_cached_instances()
+    @mutex.synchronize { @cached_nodes }
+  end
+
+  def cache_instances(instances)
+    @mutex.synchronize { @cached_nodes = instances }
```

Comment:
In my early change (read/write a `Hash` object) https://git.musta.ch/airbnb/quilt/blob/master/lib/quilt/nerve/watcher.rb#L87-L93, @nelgau mentioned zk client also has its own thread to invoke the watch callback, so concurrent read/write is possible. For my previous case, inserting new data to `Hash` object may cause internal array to resize, so even with GVL, the we cannot guarantee the consistent read/write.

But for this particular case, I think Ben made great point. If the object assignment (which I don't know if it's true in ruby) read/write `@cached_nodes` without mutex should be safe.